manage-todos-with-realtime-agent/realtime_over_websockets/main.py

```diff
@@ -0,0 +1,183 @@
+
+    audio_adapter = WebSocketAudioAdapter(websocket, logger=logger)
+
+    system_message = """You are an AI voice assistant powered by AG2 and the OpenAI Realtime API.
+You will help user manage todos. You can add, modify, and delete todos.
+
+- When the user say something like "I need to do ...", call the add_todo function to add a new todo.
+- When the user say something like "I have done ...", check if the todo is in the list and mark it as done using `modify_todo`. If not found, the assistant will add a new todo and mark it as done using `add_todo`.
+- Help user modify or delete a todo as needed with `modify_todo` and `delete_todo` functions. Understand what todo the user is referring and get the correct todo.
+- The task id should be assigned automatically by you, each task should have a unique id. The id starts from 1. You don't need to tell the user about task id.
+"""
+    system_message += "\nCurrent date and time: " + time.strftime(
+        "%Y-%m-%d %H:%M:%S", time.localtime()
+    )
+    data = get_data()
+    if len(data) > 0:
+        system_message += f"\n\nTodos from previous session:\n{format_todo_str(data)}"
+    else:
+        system_message += "\n\nNo todos from previous session."
+
+    realtime_agent = RealtimeAgent(
+        name="Weather Bot",
+        system_message=system_message,
+        llm_config=realtime_llm_config,
+        audio_adapter=audio_adapter,
+        logger=logger,
+    )
+
+    @realtime_agent.register_realtime_function(  # type: ignore [misc]
+        name="add_todo", description="Add a todo"
+    )
+    def add_todo(
+        id: Annotated[int, "task id"],
+        task: Annotated[str, "task description"],
+        status: Annotated[str, "task status, either done or open"],
+    ):
+        # if exist, load the data
+        # if not exist, create empty list
+        data = get_data()
+        data.append(
+            {
+                "id": id,
+                "task": task,
+                "status": status,
+            }
+        )
+        save_data(data)
+        return f"Todo with id {id} added.\n" + format_todo_str(data)
+
+    @realtime_agent.register_realtime_function(  # type: ignore [misc]
+        name="modify_todo", description="Modify a todo"
+    )
+    def modify_todo(
+        id: Annotated[int, "task id"],
+        task: Annotated[str, "task description"],
+        status: Annotated[str, "task status, either done or open"],
+    ):
+        data = get_data()
+        is_todo_found = False
+        for todo in data:
+            if todo["id"] == id:
+                todo["task"] = task
+                todo["status"] = status
+                is_todo_found = True
+                break
+
+        if not is_todo_found:
+            msg = f"Todo with id {id} not found."
+            add_msg = add_todo(id, task, status)
+            return msg + "\n" + add_msg
+
+        save_data(data)
+        return f"Todo with id {id} updated.\n" + format_todo_str(data)
```

Comment:
Use the modify_todo from todo_utils

```suggestion
    @realtime_agent.register_realtime_function(  # type: ignore [misc]
        name="modify_todo", description="Modify a todo"
    )(modify_todo)
```